### ISSUE TITLE
Type-checking crash with top-level binding and an expression using it along with a lambda or case binding the same name

### ISSUE THREAD

**Issue by jvoigtlaender** [elm-lang/elm-compiler#1425]
The below was investigated with Elm 0.17 at http://elm-lang.org/try.

This program works as expected:

``` elm
import Html

name =
    "Whatever"

main =
    always (Html.text name) (\_ -> "Doesn't matter")
```

But this small and supposedly innocent variation of it (naming the parameter of the lambda the same as the top-level `name` that is also used elsewhere in `main`) crashes the compiler:

``` elm
import Html

name =
    "Whatever"

main =
    always (Html.text name) (\name -> "Doesn't matter")
```

The error message is:

```
elm-make: Could not find `name` when solving type constraints.
elm-make: thread blocked indefinitely in an MVar operation
```

The `name` being a top-level binding is a precondition for getting the error. If one puts it into a `let` inside `main` instead, the error does not occur.

The error is not specific to `main`, and it also occurs for `case`\- instead of lambda-binding:

``` elm
import Html

name =
    "Whatever"

main =
    Html.text "Test"

other =
    always (Html.text name) (case 42 of name -> "Doesn't matter")
```

So the conditions for the error here are:
- define a top-level binding (constant or function) and
- use it in an expression
- that also contains (separately) a lambda or `case` that binds the same name.

Edit: Revised to incorporate additional findings from exchange with @mgold.

**Comment by mgold** [comment]
You have incorrectly identified the root cause. The problem is not that fields may not be assigned to constants of the same name. As proof, simply change every occurrence of `model` except the field name to something else, and see that the tests still fail:

```elm
import Html exposing (..)
import Html.App exposing (..)

m =
    "Test"

main =
    beginnerProgram
        { model = m
        , view = \m -> div [] [ text m ]
        , update = \_ m -> m
        }
```

Instead, the problem is that the lambdas also bind `model`. This succeeds:

```elm
model =
    "Test"

main =
    beginnerProgram
        { model = model
        , view = \m -> div [] [ text m ]
        , update = \_ m -> m
        }
```

Interestingly, the conflicting name must be a top-level constant, as this compiles:

```elm
main =
  let
    m = "Test"
  in
    beginnerProgram
        { model = m
        , view = \m -> div [] [ text m ]
        , update = \_ m -> m
        }
```

All of this testing was done on elm-lang/try, so it's possible local compilation will yield a different result.

To summarize, to reproduce you must
* define a **top level constant** and
* assign it as the **value of a field in a record**
* that also has fields that are **lambdas that bind the name of the constant**.

**Comment by jvoigtlaender** [comment]
@mgold, thanks for investigating! In compliance with how Evan wants to see issues, I'll incorporate your findings into the opening comment above.

**Comment by jvoigtlaender** [comment]
@mgold, see my revised comment above. Does it incorporate all your findings?

**Comment by jvoigtlaender** [comment]
@mgold, I simplified the example further. Records are not needed. I think all the comments except the current first one should be deleted for the sake of keeping the issue clean.

**Comment by mgold** [comment]
I've deleted my large comment above. I'm curious whether there is something special about `main` that causes the issue to occur or if it can happen in any function.

**Comment by jvoigtlaender** [comment]
It's not specific to `main`. Moreover, it also occurs with `case` instead of lambda. I've revised the opening comment further with that information.

**Comment by jvoigtlaender** [comment]
Also, turns out there was a previous issue observing a special case of this: https://github.com/elm-lang/elm-compiler/issues/952.

**Comment by Mouvedia** [comment]
This issue should be added to #1373.

**Comment by rtfeldman** [comment]
Seems to be fixed in 0.18. Thanks for reporting this, @jvoigtlaender!

<img width="741" alt="screen shot 2017-01-07 at 9 23 37 pm" src="https://cloud.githubusercontent.com/assets/1094080/21747585/9d59b8be-d51f-11e6-9395-aa88a34ceed3.png">
